### PR TITLE
Preb 21 q multi imp support

### PR DESCRIPTION
**Pull request by jgresens**
> For internal review

## Type of change
- [x] Feature

## Description of change
Smaatos Ad Server doesn't fully supports multiple impressions per request, creating an individual request per impression is a simple but effective workaround.

- contact email of the adapter’s maintainer: prebid@smaato.com
- [x] official adapter submission

### CODE REVIEW

File: modules/smaatoBidAdapter.js [hunk 1]
```diff
@@ -8,42 +8,62 @@ const SMAATO_ENDPOINT = 'https://prebid.ad.smaato.net/oapi/prebid';
 const CLIENT = 'prebid_js_$prebid.version$_1.1'
```
Comment (Enigo) [inline modules/smaatoBidAdapter.js:8]: please increase the version, 1.2 should do

File: modules/smaatoBidAdapter.js [hunk 2]
```diff
@@ -61,95 +81,67 @@ const buildOpenRtbBidRequestPayload = (validBidRequests, bidderRequest) => {
...
   }
 
-  const eids = utils.deepAccess(validBidRequests[0], 'userIdAsEids');
-  if (eids && eids.length) {
-    utils.deepSetValue(request, 'user.ext.eids', eids);
+  let payload = {};
+  if (validBidRequests.length === 1) {
+    payload = buildOpenRtbPayloadForIndividualBidRequest(validBidRequests[0])
+  } else {
+    payload = validBidRequests.map(buildOpenRtbPayloadForIndividualBidRequest)
```
Comment (Enigo) [inline modules/smaatoBidAdapter.js:116]: why can't we do just
`payload = validBidRequests.map(buildOpenRtbPayloadForIndividualBidRequest)`
for both cases? if it is one or multiple - shouldn't matter in this case

File: test/spec/modules/smaatoBidAdapter_spec.js [hunk 3]
```diff
@@ -368,222 +102,537 @@ describe('smaatoBidAdapterTest', () => {
...
+        const bidRequest1BidId = '1111'
+        utils.deepSetValue(bidRequest1, 'bidId', bidRequest1BidId);
+        const bidRequest2 = utils.deepClone(singleBannerBidRequest);
+        const bidRequest2BidId = '2222'
+        utils.deepSetValue(bidRequest2, 'bidId', bidRequest2BidId);
 
-    it('sends usp if exists', () => {
-      expect(this.req.regs.ext.us_privacy).to.equal('uspConsentString');
-    });
+        const reqs = JSON.parse(spec.buildRequests([bidRequest1, bidRequest2], defaultBidderRequest).data)
```
Comment (Enigo) [inline test/spec/modules/smaatoBidAdapter_spec.js:235]: if I understood it all correctly this is not multi-imp, it is two separate bid requests.

we should have one bidRequest with two imp inside

File: test/spec/modules/smaatoBidAdapter_spec.js [hunk 4]
```diff
@@ -368,222 +102,537 @@ describe('smaatoBidAdapterTest', () => {
...
+            mediaType: 'banner'
+          }
+        }
+      ]);
     });
+
     it('single video reponse', () => {
-      const bids = spec.interpretResponse(openRtbBidResponse(ADTYPE_VIDEO), request);
-      assert.deepStrictEqual(bids, interpretedBidsVideo);
+      const bids = spec.interpretResponse(buildOpenRtbBidResponse(ADTYPE_VIDEO), request);
```
Comment (Enigo) [inline test/spec/modules/smaatoBidAdapter_spec.js:595]: I'd say we need a case for multi-imp request here as well, just to know what is returned

File: test/spec/modules/smaatoBidAdapter_spec.js (hunk 3 above)
Comment (jgresens) [reply test/spec/modules/smaatoBidAdapter_spec.js:235]: yeah you are right, completely overlooked that

File: test/spec/modules/smaatoBidAdapter_spec.js (hunk 4 above)
Comment (jgresens) [reply test/spec/modules/smaatoBidAdapter_spec.js:618]: This is for interpreting the response from our server though :confused:

Comment (Enigo) [reply test/spec/modules/smaatoBidAdapter_spec.js:618]: ah okay

**Comment by Enigo** [comment]
lgtm

**Pull request closed without merging**